src/enclave/rpccontext.h

```diff
@@ -0,0 +1,289 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "node/clientsignatures.h"
+#include "node/entities.h"
+#include "node/rpc/jsonrpc.h"
+
+#include <variant>
+#include <vector>
+
+namespace enclave
+{
+  static constexpr size_t InvalidSessionId = std::numeric_limits<size_t>::max();
+
+  struct SessionContext
+  {
+    size_t client_session_id = InvalidSessionId;
+    std::vector<uint8_t> caller_cert = {};
+
+    //
+    // Only set in the case of a forwarded RPC
+    //
+    struct Forwarded
+    {
+      // Initialised when forwarded context is created
+      const size_t client_session_id;
+      const ccf::CallerId caller_id;
+
+      Forwarded(size_t client_session_id_, ccf::CallerId caller_id_) :
+        client_session_id(client_session_id_),
+        caller_id(caller_id_)
+      {}
+    };
+    std::optional<Forwarded> fwd = std::nullopt;
+
+    // Constructor used for non-forwarded RPC
+    SessionContext(
+      size_t client_session_id_, const std::vector<uint8_t>& caller_cert_) :
+      client_session_id(client_session_id_),
+      caller_cert(caller_cert_)
+    {}
+
+    // Constructor used for forwarded and PBFT RPC
+    SessionContext(
+      size_t fwd_session_id_,
+      ccf::CallerId caller_id_,
+      const std::vector<uint8_t>& caller_cert_ = {}) :
+      fwd(std::make_optional<Forwarded>(fwd_session_id_, caller_id_)),
+      caller_cert(caller_cert_)
+    {}
+  };
+
+  struct ErrorDetails
+  {
+    int code;
+    std::string msg;
+  };
+
+  struct RpcResponse
+  {
+    std::variant<ErrorDetails, nlohmann::json> result;
+  };
+
+  class RpcContext
```

Comment:
Minor: it seems that we've got some inconsistencies between `Rpc` and `RPC` - I'm happy with either but we should agree on one.

Reply:
I think we should use `Rpc` to avoid `CCFJSONRPC`. I think I've used `Rpc` in all the new/moved things in this PR, and we can look at a unification pass in future if we don't reach consensus more naturally.